File: frontend/src/entities/accounts-payable/api/accounts-payable.mapper.ts
/**
 * AccountsPayable mapper - transforms API responses to domain models.
 * INTERNAL: Not exported from entity public API.
 */

import type { VendorPaymentMethod } from './record-payment';
import type { CalculatedAPStatus } from '../model/accounts-payable-status';
import type { AccountsPayable, AccountsPayableDetail, APAgingSummary } from '../model/accounts-payable';
import type { VendorPayment } from '../model/vendor-payment';

/**
 * API response type for accounts payable.
 */
export interface AccountsPayableResponse {
  id: number;
  // Disbursement cause fields
  causeType: string;
  causeId: number;
  causeReferenceNumber: string;
  // Vendor info
  vendorId: number;
  vendorName: string;
  totalAmount: number;
  currency: string;
  dueDate: string | null;
  notes: string | null;
  createdAt: string;
  // Calculated fields
  totalPaid: number;
  remainingBalance: number;
  isOverdue: boolean;
  daysOverdue: number;
  agingBucket: string;
  calculatedStatus: CalculatedAPStatus;
}

/**
 * API response type for AP aging summary.
 */
export interface APAgingSummaryResponse {
  agingBucket: string;
  count: number;
  totalAmount: number;
  totalPaid: number;
  remainingBalance: number;
}

/**
 * API response type for vendor payment.
 */
export interface VendorPaymentResponse {
  id: number;
  accountsPayableId: number;
  paymentDate: string;
  amount: number;
  paymentMethod: VendorPaymentMethod;
  referenceNumber: string | null;
  notes: string | null;
  recordedById: number;
  recordedByName: string;
  createdAt: string;
}

/**
 * API response type for accounts payable detail with payments.
 */
export interface AccountsPayableDetailResponse extends AccountsPayableResponse {
  payments: VendorPaymentResponse[];
}

/**
 * Map API response to AccountsPayable domain model.
 */
export function mapToAccountsPayable(response: AccountsPayableResponse): AccountsPayable {
  return {
    id: response.id,
    causeType: response.causeType,
    causeId: response.causeId,
    causeReferenceNumber: response.causeReferenceNumber,
    vendorId: response.vendorId,
    vendorName: response.vendorName,
    totalAmount: response.totalAmount,
    currency: response.currency,
    dueDate: response.dueDate,
    notes: response.notes,
    createdAt: response.createdAt,
    totalPaid: response.totalPaid,
    remainingBalance: response.remainingBalance,
    isOverdue: response.isOverdue,
    daysOverdue: response.daysOverdue,
    agingBucket: response.agingBucket,
    calculatedStatus: response.calculatedStatus,
  };
}

/**
 * Map API response to APAgingSummary domain model.
 */
export function mapToAPAgingSummary(response: APAgingSummaryResponse): APAgingSummary {
  return {
    agingBucket: response.agingBucket,
    count: response.count,
    totalAmount: response.totalAmount,
    totalPaid: response.totalPaid,
    remainingBalance: response.remainingBalance,
  };
}

/**
 * Map API response to VendorPayment domain model.
 */
export function mapToVendorPayment(response: VendorPaymentResponse): VendorPayment {
  return {
    id: response.id,
    accountsPayableId: response.accountsPayableId,
    paymentDate: response.paymentDate,
    amount: response.amount,
    paymentMethod: response.paymentMethod,
    referenceNumber: response.referenceNumber,
    notes: response.notes,
    recordedById: response.recordedById,
    recordedByName: response.recordedByName,
    createdAt: response.createdAt,
  };
}

/**
 * Map API response to AccountsPayableDetail domain model.
 */
export function mapToAccountsPayableDetail(response: AccountsPayableDetailResponse): AccountsPayableDetail {
  return {
    id: response.id,
    causeType: response.causeType,
    causeId: response.causeId,
    causeReferenceNumber: response.causeReferenceNumber,
    vendorId: response.vendorId,
    vendorName: response.vendorName,
    totalAmount: response.totalAmount,
    currency: response.currency,
    dueDate: response.dueDate,
    notes: response.notes,
    createdAt: response.createdAt,
    totalPaid: response.totalPaid,
    remainingBalance: response.remainingBalance,
    isOverdue: response.isOverdue,
    daysOverdue: response.daysOverdue,
    agingBucket: response.agingBucket,
    calculatedStatus: response.calculatedStatus,
    payments: response.payments.map(mapToVendorPayment),
  };
}

File: backend/src/main/java/com/wellkorea/backend/core/finance/infrastructure/mapper/AccountsPayableMapper.java
package com.wellkorea.backend.core.finance.infrastructure.mapper;

import com.wellkorea.backend.core.finance.api.dto.query.AccountsPayableDetailView;
import com.wellkorea.backend.core.finance.api.dto.query.AccountsPayableSummaryView;
import com.wellkorea.backend.core.finance.api.dto.query.VendorPaymentView;
import org.apache.ibatis.annotations.Mapper;
import org.apache.ibatis.annotations.Param;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

/**
 * MyBatis mapper for AccountsPayable queries with calculated status.
 * Status is computed from vendor_payments table, not stored in AP.
 */
@Mapper
public interface AccountsPayableMapper {

    /**
     * Find AP detail by ID with calculated payment status.
     */
    Optional<AccountsPayableSummaryView> findDetailById(@Param("id") Long id);

    /**
     * Find APs with filters and calculated status.
     * Status is calculated from vendor_payments.
     *
     * @param vendorId         optional vendor ID filter
     * @param causeType        optional disbursement cause type filter (PURCHASE_ORDER, EXPENSE_REPORT, etc.)
     * @param calculatedStatus optional calculated status filter (PENDING, PARTIALLY_PAID, PAID)
     * @param overdueOnly      if true, only return overdue APs
     * @param dueDateFrom      optional due date range start filter
     * @param dueDateTo        optional due date range end filter
     * @param limit            pagination limit
     * @param offset           pagination offset
     * @return list of AP summary views
     */
    List<AccountsPayableSummaryView> findWithFilters(
            @Param("vendorId") Long vendorId,
            @Param("causeType") String causeType,
            @Param("calculatedStatus") String calculatedStatus,
            @Param("overdueOnly") Boolean overdueOnly,
            @Param("dueDateFrom") LocalDate dueDateFrom,
            @Param("dueDateTo") LocalDate dueDateTo,
            @Param("limit") int limit,
            @Param("offset") long offset
    );

    /**
     * Count APs with filters.
     *
     * @param vendorId         optional vendor ID filter
     * @param causeType        optional disbursement cause type filter
     * @param calculatedStatus optional calculated status filter
     * @param overdueOnly      if true, only count overdue APs
     * @param dueDateFrom      optional due date range start filter
     * @param dueDateTo        optional due date range end filter
     * @return count of matching APs
     */
    long countWithFilters(
            @Param("vendorId") Long vendorId,
            @Param("causeType") String causeType,
            @Param("calculatedStatus") String calculatedStatus,
            @Param("overdueOnly") Boolean overdueOnly,
            @Param("dueDateFrom") LocalDate dueDateFrom,
            @Param("dueDateTo") LocalDate dueDateTo
    );

    /**
     * Get AP aging summary (grouped by aging bucket).
     */
    List<APAgingSummary> getAgingSummary();

    /**
     * Find AP detail by ID with payment history.
     *
     * @param id AP ID
     * @return AP detail with payments
     */
    Optional<AccountsPayableDetailView> findDetailWithPaymentsById(@Param("id") Long id);

    /**
     * Find payments by accounts payable ID.
     * Used as nested select for detail view.
     *
     * @param accountsPayableId the AP ID
     * @return list of vendor payment views
     */
    List<VendorPaymentView> findPaymentsByAccountsPayableId(@Param("accountsPayableId") Long accountsPayableId);

    /**
     * AP aging summary record.
     */
    record APAgingSummary(
            String agingBucket,
            int count,
            java.math.BigDecimal totalAmount,
            java.math.BigDecimal totalPaid,
            java.math.BigDecimal remainingBalance
    ) {
    }
}
